src/shogun/multiclass/KNNSolver.h

```diff
@@ -83,9 +84,15 @@ class KNNSolver : public DistanceMachine
 		 */
 		 virtual SGVector<int32_t> classify_objects_k(std::shared_ptr<Distance> d, const int32_t num_lab, SGVector<int32_t>& train_lab, SGVector<int32_t>& classes) const = 0;
 
+		virtual bool train_KNN(std::shared_ptr<Distance> knn_distance) const;
+
+		virtual bool compute_nearest_neighbours();
```

Comment:
The linker error you are getting is because you declared these functions but never defined them. You just have the definition in the derived class right? In that case these should be pure virtual functions (whether that is good design, I don’t know). You could also CRTP instead.